features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/resources/ResourceDataStoreManager.kt

```diff
@@ -0,0 +1,158 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.resources
+
+import android.text.format.DateUtils
+import com.datadog.android.api.feature.FeatureSdkCore
+import com.datadog.android.api.storage.datastore.DataStoreReadCallback
+import com.datadog.android.api.storage.datastore.DataStoreWriteCallback
+import com.datadog.android.core.internal.persistence.Deserializer
+import com.datadog.android.core.persistence.Serializer
+import com.datadog.android.core.persistence.datastore.DataStoreContent
+import com.datadog.android.sessionreplay.internal.ResourcesFeature.Companion.SESSION_REPLAY_RESOURCES_FEATURE_NAME
+import com.datadog.android.sessionreplay.model.ResourceHashesEntry
+import java.util.Collections
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicLong
+
+internal class ResourceDataStoreManager(
+    private val featureSdkCore: FeatureSdkCore,
+    private val resourceHashesSerializer: Serializer<ResourceHashesEntry>,
+    private val resourceHashesDeserializer: Deserializer<String, ResourceHashesEntry>
+) {
+    private val knownResources = Collections.synchronizedSet(mutableSetOf<String>())
+    private var storedLastUpdateDateNs = AtomicLong(System.nanoTime())
+    private var isInitialized = AtomicBoolean(false) // has init finished executing its async actions
+
+    init {
+        val onDeleteDatastoreCallback = object : DataStoreWriteCallback {
+            override fun onSuccess() {
+                finishedInitializingManager()
+            }
+
+            override fun onFailure() {
+                finishedInitializingManager()
+            }
+        }
+
+        fetchStoredResourceHashes(
+            onFetchSuccessful = lambda@{ storedEntry ->
+                val storedData = storedEntry?.data
+
+                if (storedData == null) {
+                    finishedInitializingManager()
+                    return@lambda
+                }
+
+                val lastUpdateDateNs = storedData.lastUpdateDateNs
+                val storedHashes = storedData.resourceHashes
+
+                storedLastUpdateDateNs.set(lastUpdateDateNs.toLong())
+
+                if (didDataStoreExpire()) {
+                    handleDataStoreExpired(
+                        storedHashes = storedHashes.toSet(),
+                        callback = onDeleteDatastoreCallback
+                    )
+                } else {
+                    knownResources.addAll(storedHashes)
+                    finishedInitializingManager()
+                }
+            },
+            onFetchFailure = {
+                deleteStoredHashesEntry(onDeleteDatastoreCallback)
+            }
```

Comment:
I think there is unnecessary complexity here: probably we don't need to do anything for both `onFetchFailure` and `didDataStoreExpire` cases. That is because once new hashes come, we can write them to the store and this entry will override the previous one (which cannot be fetched or outdated). The only case when it won't happen - version difference (by path, I believe?), but I'm not sure it is critical.

This complexity forced to add callbacks for the `removeValue`/`setValue` APIs, which are not normally necessary (and also provide no-op value in `writeResourcesToStore`).

Reply:
What I'm concerned about is that without having this callback we need to return that the manager has finished initialization when we make the deletion call - but this call is async. So we could have a situation where a call is made to store a resourceHash while this removal of the entry is still in progress

Reply:
I didn't notice that you wrote this about onFailure. Yes, you're right - even for a corrupt entry the next write will overwrite it so this is redundant 👍  I'll remove it